keps/sig-cluster-lifecycle/kubeadm/20190722-Advanced-configurations-with-kubeadm-(Kustomize).md

```diff
@@ -0,0 +1,292 @@
+3. Kubeadm is responsible for coordinating the execution of Kustomize within the
+   init/join/upgrade workflows
+4. as a consequence of the previous point, higher-level tools/users are not
+   requested to take care of defining `kustomization.yaml` files nor to define
+   a local folder structure.
+
+Additionally, in order to simplify the first implementation of this KEP, this 
+proposal is going to assume that Kustomize patches for kubeadm are always defined
+specifically for the node where kubeadm is going to be executed.
+
+This point could be reconsidered in the future, by e.g. introducing cluster-wide
+patches and/or patches for a subset of nodes.
+
+#### Providing and storing Kustomize patches to kubeadm
+
+Before kubeadm init, Kustomize patches should be eventually provided to kubeadm
+by higher-level tools/users; patches should be provided in a well know location
+like e.g. `/etc/kubernetes/kustomize` or in a custom location specified with a CLI flag. 
+Eventually, in the future, this flag will be added to the kubeadm config API as well.
+
+In order to simplify the first implementation of this KEP, this proposal is assuming
+to use the same approach also for kubeadm join; this point could be reconsidered
+in the future, by e.g. defining a method for allowing higher-level tools/users to
+define Kustomize patches using a new CRD.
+
+#### Storing and retrieving Kustomize patches for kubeadm
+
+Kustomize patches, should be preserved during the whole cluster lifecycle, mainly
+for allowing kubeadm to preserve changes during the kubeadm upgrade workflow.
+
+In order to simplify the first implementation of this KEP, this proposal is assuming
+that Kustomize patches will remain stored in the `/etc/kubernetes/kustomize` or in a
+custom location specified with a CLI flag for the necessary time; this point could
+be reconsidered in the future, by e.g. defining a method for allowing higher-level
+tools/users to define Kustomize patches using a new CRD.
+
+### Risks and Mitigations
+
+_Confusion between kubeadm config API and kustomize_
+Kubeadm already offers a way to implement cluster settings, that is the kubeadm API
+and component configs. Adding a new feature for supporting “advanced configurations”
+can create confusion in the users.
+
+kubeadm maintainers should take care of making differences cristal clear in release notes
+and feature announcement:
+
+- The config API is well suited for most common use cases and it will remain the
+  recommended way to define core settings for a cluster, while “advanced configurations”
+  are designed for less common use cases.
+- The config API implicitly defines the main cluster variants the kubeadm team is
+  committed to support and monitor in the Kubernetes test grid, while instead higher-level
+  tools/user are responsible for the security, conformance, consistency, of 
+  “advanced configurations” for control-plane or kubelet settings. 
+
+_To not provide expected flexibility_
```

Comment:
i think this title needs to define the problem.

i'd rename it to:
> Misleading expectations on the level of flexibility

or something similar.

Reply:
done